src/odemis/acq/align/z_localization.py

```diff
@@ -0,0 +1,193 @@
+    :return:
+        z_position (float): determined z position of the feature in meter
+        warning (int/None): None = No warnings
+                        1 = from fmin_cg(scipy) max number of iterations exceeded
+                        2 = from fmin_cg(scipy) gradient and/or function calls were not changing
+                        3 = from fmin_cg(scipy) NaN result encountered
+                        4 = Outputted Z position is outside the defined maximum range, output is inaccurate
+                        5 = The Gaussian fit is not precise enough, probably because the image contains too much noise
+                        6 = The Gaussian fit found a feature to big for the current feature, the size > 85%
+    :raises:
+            ModuleNotFoundError if the module psf_extractor is not found.
+            KeyError if the calibration data is incomplete and does not include all the keys
+
+    """
+    warning = None  # Set the warning level to None, no warnings
+
+    if not psf_extractor:
+        raise ModuleNotFoundError("The psf_extractor module is not found. Cannot determine the Z position")
+
+    # Fit a Gaussian to the feature in the image
+    popt, pcov = psf_extractor.extractor.fit_gaussian_2D(image, theta=numpy.rad2deg(calibration_data["feature_angle"]), epsilon=1e-6)
+    sigma_x, sigma_y = popt[2:4]
+
+    # Ensure that for a normal distribution at least 95% falls within the set tolerance for the calculated standard
+    # deviation error (this means a confidence level of 1.96 time the standard deviation above and below the mean)
+    if (fit_tol / 0.98) * sigma_x < pcov[2] or (fit_tol / 0.98) * sigma_y < pcov[3]:
+        logging.warning(f"The image contains too much noise to determine the size of the feature accurately, "
+                        f"the results may be inaccurate.\n"
+                        f"The standard deviation error for determined feature size in x is: {pcov[2]} and for y is {pcov[3]}.")
+        warning = 5
+    elif sigma_x > 0.85 * image.shape[0] or sigma_y > 0.85 * image.shape[1]:
+        logging.warning(f"The detected size of the feature in the image is too big w.r.t. the size of the image."
+                        f"Using an image with more space around the feature might solve this problem."
+                        f"The detected feature size is {sigma_x} in x and {sigma_y} in y direction, while the total image has a "
+                        f"width of {image.shape[0]} pixels and a height of {image.shape[1]} pixels."
+                        f"The feature size should be < 85% in each direction\n"
+                        f"Current results may be inaccurate.")
+        warning = 6
+
+    # Determine the z position using the shape of the features (sigma_x/sigma_y)
+    max_range = calibration_data["z_calibration_range"][1] - calibration_data["z_calibration_range"][0]
+    fine_z = numpy.arange(0, max_range, z_stack_step_size*1e9)
```

Comment:
Instead of relying on the z_stack_step_size, maybe you could just arbitrarily take 10 or 20 values between 0 and max_range (ie, using linspace)

Reply:
I would say looking at the step size is better. Otherwise the precision of your initial guess is dependent on your max_range. That feels a bit odd.
Though the 'random value" I like. Maybe I was already thinking about removing the z_stack_step_size argument. Maybe I should just hardcode it and comment on it?

Reply:
Well, my understanding of this code is that it just tries many values between 0 and max_range, and pick the one that seems to fit best. Currently it takes as many trial values as there was stacks in the calibration zstack. That's quite arbitrary too. If you hard-code the step size, it's probably going to work in the standard cases, because max_range is usually in the same order of magnitude anyway, but that could break in some extreme cases where max_range is particularly small. I think that just hard coding the number of values to try is the way to go.